synapse/handlers/room_member.py

```diff
@@ -618,6 +604,25 @@ async def update_membership(
         Raises:
             ShadowBanError if a shadow-banned requester attempts to send an invite.
         """
+        if ratelimit:
+            if action == Membership.JOIN:
+                # Only rate-limit if the user isn't already joined to the room, otherwise
+                # we'll end up blocking profile updates.
+                (
+                    current_membership,
+                    _,
+                ) = await self.store.get_local_current_membership_for_user_in_room(
+                    requester.user.to_string(),
+                    room_id,
+                )
+                if current_membership != Membership.JOIN:
+                    await self._join_rate_limiter_local.ratelimit(requester)
+                    await self._join_rate_per_room_limiter.ratelimit(
+                        requester, key=room_id, update=False
+                    )
+            elif action == Membership.INVITE:
+                await self.ratelimit_invite(requester, room_id, target.to_string())
```

Comment:
Nice consolidation here. 👍 

I wondered why we couldn't remove `ratelimit` from other spots, but we pass it into `handle_new_client_event`.

Reply:
I had the same thought (sorry, should have left a comment to this effect), and decided that I was too scared to see if the ratelimiting logic was further disentangleable.

Reply:
Completely fair, I think this is a solid improvement.